New features for [role]

GregoryLundberg opened the pull request:
This builds upon [CelticMinstrel's changes](https://github.com/wesnoth/wesnoth/commit/d4b8fec953ca81531d323774a39da90445dfd3d6) from a few days ago.

On code-reading, I see he made allowances to bypass searching the map. Also, he asked about whether a [role] would re-assign the role if a better match was found. On testing, I found, it did. As a result, I added the following new attributes for the [role] tag.

    search_map=yes|no (optional, default yes)
    reassign=yes|no (optional, default yes)

The search_map attribute determines whether units presently on the map should be considered for the role. Note that if both search_map and search_recall_list are specified as 'no' the [else] sub-tag always executes; if there also is no [else] tag the [role] tag does nothing at all. (These might be cases for wmllint to check for, someday.)

The reassign attribute determines whether, if a unit already has the role, the role should be re-assigned to a 'better' unit. Here, 'better' means either appears earlier in the listed types, or earlier in whatever the order happens to be in the list of units matching the filters.

Consider one use case: a map with a number of Mages and one Red Mage.

1) Use [role] in the normal way to assign role=MageAdvisor to the Red Mage using types="White Mage,Red Mage,Mage".

2) Advance some of the Mages to Red Mage, some to White Mage.

3) Use [role] with the same type list and reassign=no. The same Red Mage will still have the role MageAdvisor without regard to whether 'better' advisors might now be available.

4) Kill the Red Mage with the role MageAdvisor

4) Use [role] with the same type list and reassign=no. Now, there is no unit with the role, so it must be assigned. One of the White Mages will be selected.

CelticMinstrel commented:
I recall you suggesting `search_recall_list=only` to only search the recall list. My initial version included that, but then I realized that you should be able to get the same effect with `x,y=recall,recall`. Is there some other reason to add `search_map`? The only reason I can think of is really consistency of syntax (so that map and recall list are handled in the same way), and I'm not sure that's a sufficient reason.

I had also considered the issue of reassigning but had not settled on a way to handle it. Vultraz mentioned the possibility of conditional roles, though I personally disliked the idea of putting an `[if]` or similar tag containing conditions within the role (because, what's the difference between that and nesting the `[role]` within an `[if]`?). I guess adding an explicit reassign option is a valid solution, though.

GregoryLundberg commented:
My primary reason for search_map was you had the code there but the if condition was always true. I then considered 'x,y=recall,recall' but it seemed 'illiterate' and a trick. Yes, the attribute does the same thing, but the intention is more apparent. Finally, I'm not sure overloading x and y like that was ever a good idea.

What if I want a role for a unit in an area of the map and, if not found, want to recall it into that area and only recruit in [else] .. I'm not sure that's ever been done, but it sure sounds like a valid use case.

    [role]
        role=Northwest_Scout
        type="Elvish Outrider,Elvish Rider,Elvish Scout"
        side=1
        x,y=1-10,1-19 # Up near our castle
        auto_recall=yes
        auto_recall_x,auto_recall_y=1,2 # In our castle .. yipes no attribute
        reassign=no
        search_map=yes
        search_recall_list=yes
        [else]
            [unit]
                side=1
                type="Elvish Scout"
                animate=no
                role=Northwest_Scout
                x,y=1,2 # In our castle
            [/unit]
        [/else]
    [/role]

One would expect the above to work as the recall list has no map coordinates. Either use the role already assigned (anywhere on the map, maybe it was there but we moved it) or assign it to a scout in the northwest corner (maybe we have a castle there and have left the area but left a few behind), or recall a scout (yikes .. auto_recall needs coordinates .. and I hope get_recall_units' filter ignores x and y!), or recruit a scout if all else fails.

Now, macro-ize that!

GregoryLundberg commented:
Already tested.

One would expect a role to be uniquely assigned.  That is NOT the case. Even before all these changes you could easily have several units on the recall list with a role, **AND** several on the map.

Examine your Lua .. you set filter.role to nil so get_units and get_recall_units does not filter on the role, at all. They take the first match and give it the role. If, later, someone else has the role .. well .. no-see-um no-harm-um

GregoryLundberg commented:
I am not sure we even want a uniqueness requirement for roles. But my reassign=no attribute gets you there when you do.

I was also thinking .. can a unit have more than one role?  If not, why not? It does, though, means the role attribute in the unit needs to be split and merged and then we have the problem of removing a role without removing all roles and sheesh this is starting to sound like HTML and Javascript and the class= problem

CelticMinstrel commented:
> Finally, I'm not sure overloading x and y like that was ever a good idea.

I have to admit that you have a point here.

Regarding your example, I believe recalling a unit automatically places it next to the leader, unless an explicit position is given. If you need a specific position you could always set `auto_recall=no` and do an explicit `[recall]` after the `[role]`, since it never hurts to attempt to recall a unit for which there's no match.

I don't think it's really necessary to allow a unit to have multiple roles, but it _might_ be good to ensure that `[role]` does not select a unit that already has a role.

GregoryLundberg commented:
I was gonna check [recall] but the wiki is down, to me (DNS not found). It's T-storms here so I'm not worrying until the weather clears.

OH .. yes. "Gee, I just assigned Leader role the a Knight and now when I assign Lieutenant role it goes to the same Knight and I have no Leader ... whassupwidat?"

GregoryLundberg commented:
The point of auto_recall was to avoid doing a recall when we found a unit on the map, or did the [else] to recruit one. Need to think on that. After dinner.

CelticMinstrel commented:
Well, "leader" is not normally a role though.

GregoryLundberg commented:
I was considering replacing auto_recall with [recall] or [auto_recall] (like that better) and allowing options there. x,y would be the first. Dunno what else might be needed but allowing for future is always good. IF we need x and y we might as well allow for future with a sub-tag to hold em. Been working on a testbed campaign/scenario which is an "interesting" project to pull off without the wiki and having to do it from memory and what you can find in the source tree

gfgtdf commented:
> I recall you suggesting search_recall_list=only to only search the recall list. My initial version included that, but then I realized that you should be able to get the same effect with x,y=recall,recall

Wouldn't this also be aan argument against a `search_recall_list=no` key since you can just use `[not] x,y=recall,recall [not]` ?

GregoryLundberg commented:
Everything we're talking about can be done in WML without the changes. The question is just how nasty a macro you want to stream-line it for common usages. I played with several ideas and was headed toward pretty much what we have here. I'm still working my way through the mainline campaigns and am still finding cases where I think 'hmm, that's different' most can be handled by considering it a special case of something already seen, but every now and then ...

GregoryLundberg commented:
Need to re-think the reassign=no logic. Got a surprise while testing and need to investigate.

GregoryLundberg commented:
It looks like we can do

    [role]
        ...
        [else] ... [/else]
        [else] ... [/else]
        ...
    [/role]

Where each [else] block is a conditional and they are processed in order until one returns true? Or was the use of an iterator just an expedient choice? Not sure what we'd do in a series of [else] blocks that we'd not do more clearly as an [if] or [switch] inside a single [else]

Which calls the question: is [else] really the best name? Perhaps [on_no_match] or something else more literate which gets across the idea more clearly? That has the advantage of being obviously different that the existing [else] tag.

CelticMinstrel commented:
@gfgtdf I hadn't thought of that, but I'd point out that if you have xxx=yes then you must also support xxx=no even if it's possible to get the same effect in another way.

When there are multiple `[else]` blocks, all of them will be executed - it's equivalent to packing their entire contents into a single block. I only did that because there is a tendency in control flow blocks to accept multiples of the command blocks - for example, loops accept multiple `[do]` blocks, and `[object]` accepts multiple `[else]` blocks.

Speaking of `[object]`, that was also a factor in my choice of `[else]` here in `[recall]`. The use of `[else]` in `[object]` is pretty much the same as it is here - it executes if no eligible unit was found for the object. The `[object]` tag also supports a `[then]` clause, actually. (It's in wml/object.lua rather than wml-tags.lua, if you want to take a look at the code.)

CelticMinstrel commented on data/lua/wml-tags.lua:
```diff
@@ -895,6 +895,19 @@ function wml_actions.print(cfg)
+-- Helper for wml_actions.role to do the [auto_recall] child
+-- TODO enforce no StandardUnitFilter
+-- TODO enforce there can be only one [auto_recall] child
+function auto_recall(cfg, unit)
```
You should declare this function local.

GregoryLundberg commented:
Well, of course, existing practice is a good guide. Thanks.

That _`if action ~= "none" then return end`_ is what caused me to wonder why it's leaving before looking at all the [else] sub-tags.

CelticMinstrel commented:
That condition is only true if a `[break]`, `[continue]`, or `[return]` tag was executed within the current `[else]`.

CelticMinstrel commented:
Are you going to address my commit comment above?

GregoryLundberg commented:
Yeah, I spotted the missing 'local' and it's already on my list. 

The TODOs, I can handle the second, no problem, warning to stderr at debug level is good enough .. anyone testing their work should see the second [auto-recall] didn't happen or it was meaninless, anyway.

But I'm unsure how to proceed for the first TODO. Is there a function for that? It would be nice for the code here not to have to know about future standard unit filters.

GregoryLundberg commented:
I was also thinking, even though auto recall = yes has only been there a few days, it's possible that someone doing a new campaign against master might be using it. (I am but I can adjust) .. and it should not be too hard to allow either the attribute or the sub-tag if you think that's a good idea.

GregoryLundberg commented:
On a totally different track: about event name=scenario_end, this fires on debug/next level command. One place I'm looking at it is at the split in TSG. As is, it will always take the elven path, even if the player sided with the bandits. That means Ethiliel is gone and you'll auto-defeat in a bit. So I was wondering if I could do [endlevel] inside the event to set the correct next scenario. All wiki says is 'can be changed' but I don't see any way to change other then endlevel tag.

gfgtdf commented:
hmm victory events should also be fired when the scenario is ended with debug/next level command. Also lua has a wesnoth.set_next_scenario function that doesn't end the scenario (it just changes [scenario]next_scenario= attribute), maybe that can be used when the player decides his alliances

GregoryLundberg commented:
Yes, victory fires on debug/next. No, you cannot change the next scenario any way other then endlevel. Was thinking about seeing if lua can do it.

Tested and victory/defeat fires before session end but there is no way to know if they were normal endlevel or debug next level. Didn't test choose level.

At the split in TSG the only way to go with the bandits is with choose level. And whichever way you go, if the scenario was not ready for you things will go badly. If session end cannot CHANGE the next session, if really SHOULD be able to read the value and ensure everything is copacetic on the way out the door no matter how we got headed for that door.

CelticMinstrel commented:
> I was also thinking, even though auto recall = yes has only been there a few days, it's possible that someone doing a new campaign against master might be using it.

It's possible, but such a person should understand that they're working with an unstable version which could change at any time without warning. You really shouldn't be developing campaigns on master - if you want to use 1.13, you should use the latest release (currently 1.13.4).

As for whether we should support both, there's no need to provide compatibility support for features that did not make it into a release.

> But I'm unsure how to proceed for the first TODO. Is there a function for that? It would be nice for the code here not to have to know about future standard unit filters.

A function for what? You're trying to prevent them from using an SUF, right? So all you have to do is copy the attributes that aren't part of the SUF into a new table and pass that to recall.

By the way, about the use of `end scenario` events, we intend to revert most of them (probably excluding LoW) since the motivation for switching to them was highly dubious.

@gfgtdf - Don't put it in `game_config` or `current`. It should go in a new `level` table (though I suppose the `level` table itself could go in `current`, but I'm not sure that's the best idea). Or `wesnoth.next_scenario` would be fine with me, too. Note that the `level` table (if present) would also contain the scenario ID and some other WML attributes from the `[scenario]` tag.

gfgtdf commented:
> No, you cannot change the next scenario any way other then endlevel

Hmm not sure if i understood you here correctly, but i just said this exactly this can be done with wesnoth.set_next_scenario

GregoryLundberg commented:
The changes to the scenario WML (victory to scenarion_end) did look more like caught in a dragnet than arrested for cause and probably should be reverted on the principle of least-change.

The feature of session_end, however is much nicer than simply providing a way to sync changes. If I were doing it, I'd suggest the following general flow:

[session]
[event]=prestart
[event]=start
...
[event]=session_ending (allows changes to next session and can cancel the end session)
[event]=victory/defeat (local end handling)
[event]=session_end (networked end handling, cannot change but can read and react)

gfgtdf commented:
I dont really see how the victory -> scenario_end changes relate to the issue in TSG, specially since i didnt any changes in TSG.

GregoryLundberg commented:
You didn't .. I'm playing with the code there.

I find the "last chance"-ness of session_end very handy for cleaning up the mess for far more than just sync'ing network play.

It's a case of unintended consequences. You needed it as a common sync'able end point. But as a common point at the end it has a lot of usefulness, even if we ignore the network effects.

GregoryLundberg commented:
Frankly, TSG is a dog pile and screams for a re-write. But I'm restraining myself and doing only the bare minimum needed to fix the bugs. Sometimes that reduces the size of the dog pile but, usually, it's just cleaning up the boundaries so the smell stays local.

gfgtdf commented:
> It's possible, but such a person should understand that they're working with an unstable version which could change at any time without warning. You really shouldn't be developing campaigns on master - if you want to use 1.13, you should use the latest release (currently 1.13.4).

I don't say that we need to care about compability with unreleased versions, but i still disagree here, specailly when you are part of the dev team, its usually much easier to develop for latest master than for 1.13.4 or someting, you can fix core bugs that you notice while develpoing immidiateley and you can also implement core features that you need both without having to delay your campaign developement until the next release, specailly the bugfixing part is also valuable for the core wesnoth game. And for the compabiltiy, since its usually rather small changes its not really that much work for adapt them. Also its often possibel to estimate how stable a new fewture if (how likeley it is that it will be changed)

CelticMinstrel commented:
> Frankly, TSG is a dog pile and screams for a re-write.

I seem to recall liking it when I played it, though that was quite awhile ago (in wesnoth 1.10).

GregoryLundberg commented:
TSG is very playable, especially for a new player, and especially for new players who take the elven choice (the bandit choice end-game looks very intimidating!).

The problem is when an old-hand comes back to it. I, for instance, can quite easily win S01 without getting Sir Gerrick. He's so weak, you never miss him. But, at some point, a few scenes later, he's going to become the leader and, at that point, it's good night Irene, don't bother locking up, we're outta here.

My fix for that (and again in S02 and S03) is to use session_end to check and if you don't have a required NPC put it into your recall list. Game-play and on-screen remain unchanged but you're not longer left in a lose/crash situation a few scenarios later.

GregoryLundberg commented:
When I get to cleaning up the split at S05, while I can fix up the NPC  roster for the split, since I don't know where we're going, I can only go by the data. If we're actually headed elsewhere (choose level probably) I can't fix for that.

GregoryLundberg commented:
I did it! I got Debug/Nest fully working on TSG S05. Looking at lua/wml/endlevel I found how to use Lua to set the next scenario. So, a little if'n to tell if I need to ask where to go, and do the cleanup needed for going there, and viola!

OK, back to the issue for this PR ...

@CelticMinstrel I did the two things I am sure of. Still looking at how to disallow SUF stuff in [auto_recall]. My issue is I either need a full list of SUF to remove or I need a full list of [recall] w/o SUF to keep. Either way I'm making a second place you need to go to add an attribute to remove/copy and that's a long-term maintenance headache. Still looking into it.

CelticMinstrel commented:
You know, when I said "make this function local", I was just thinking "stick the local keyword in front" - you didn't have to move it. That said, it probably doesn't hurt to make it local to the recall function, since that's the only place it's actually used. (Making it local in that way actually means you could remove the arguments, since they're the same for both calls and would be captured by the closure anyway.)

By the way, you seem to be consistently calling that event "session_end" instead of "scenario_end" for some reason?

CelticMinstrel commented:
Oh right, about disallowing SUFs - the basic idea I was thinking of is like this:

```lua
local recall_cfg = {}
-- Copy over any allowed attributes
recall_cfg.x = cfg.x
recall_cfg.y = cfg.y
-- And pass the result to the WML tag instead of the original config
wml_actions.recall(recall_cfg)
```

It has the disadvantage of requiring manual update if `[recall]` every gains any new keys that you would want to support in `[auto_recall]`, but I don't think there's any situation that'll be perfect in light of changes to both SUF and `[recall]`.

GregoryLundberg commented:
I'll got with remove all but [recall]. And a comment in the [recall] C++ warning of it might be a good idea.

CelticMinstrel commented:
Huh? There's no `[recall]` within `[role]`, right?

GregoryLundberg commented:
What I mean is I'll keep all [recall] attributes inside an [auto-recall] and not copy any others, into a new objects for passing to wesnoth...recall function. As opposed to scanning for SUF to remove.

CelticMinstrel commented:
I believe the config parameter you receive is immutable anyway (because it's a vconfig). That means you'd need to parse it (ie `cfg.__parsed` or `helper.parsed(cfg)`) before you can delete anything. And there are tons of SUF attributes and tags.

GregoryLundberg commented:
I think this is now done. I've tested it (lightly) and everything seems OK. I can design a more rigorous test plan if warranted.

CelticMinstrel commented:
The SUF attributes are implemented in a totally different file, so if they add attributes at that location, it's definitely `[recall]`-specific! :P Anyway, thanks for that comment.

GregoryLundberg commented:
Wow. Reassign=no is going to be **_real_** popular.

GregoryLundberg commented:
Replaces with a consolidated patch. Much easier to read. And, with auto-recall=yes having been committed, I'm going to need to add patches to change to the sub-tag. So I wanted to keep the patch list short and clear.

CelticMinstrel commented:
You misspelled auto_recall in your commit message. Not major problem, but if you want to fix it...

GregoryLundberg commented:
Fixed.

If nobody has anything else, I think it's ready.

CelticMinstrel commented:
I thought @vultraz had some comments, but it seems he didn't mention anything after all...?

GregoryLundberg commented:
Maybe on another thread? I do recall he had some opinions. I'd suggest waiting a while to be sure, or contact him. I'd really like to finalize the [role] changes and not have another round for a while, if at all possible.

What we have, so far, is great for handling the edge cases which were generally ignored. It should make many of the mainline campaigns much more resilient. I hope the UMC creators twig to it.

CelticMinstrel commented:
I'll wait awhile. He should be notified by the mention here, but I'll also poke him if I see him in IRC.

Vultraz commented:
My critique was I thought that instead of `search_recall` and `search_map`, there should be a tristate `search=map/recall/all` key.

CelticMinstrel commented:
I'll note that that leaves `search=all` potentially unclear at first glance.

GregoryLundberg commented:
Well, search_map=no with search_recall_list=no means what .. lessee ...

If there is a unit with the role, and reassign=no, then do nothing. Otherwise, run the [else] sub-tag(s).

So, if [else] is recruit, it's "If I have a unit with the role, fine, otherwise recruit" or it's a bloated [unit] tag

Ignoring the bloated [unit] case, having the fourth state (search=no/map/recall/map_and_recall) actually does actually do something useful. It's [if][not][have_unit][then][unit] (sic).

GregoryLundberg commented:
So, which do we find more literate?

    search_map=yes|no            default yes
    search_recall_list=yes|no    default yes

or

    search=no|map|recall_list|map_and_recall_list    default map_and_recall_list

CelticMinstrel commented:
There's also the option that came up (I think it was your idea?) in the earlier thread - `search_recall_list=no|yes|only`. I don't have a strong objection to the current way either, though.

I don't think we need to support a `[role]` which searches neither the map nor the recall list.

GregoryLundberg commented:
Well, it came along for the ride with these changes. My last patch bails out early if there is nothing to do; but does allow an [else]-only role to run and execute the [else] block(s).

But, I agree.

Which means it's either

    search=map|recall|map_and_recall    optional, default map_and_recall

or

    search_recall_list=yes|no|only    optional, default=yes

Since search_recall_list is analogous to [have_unit] I will go with that. Give me a while (cooking dinner) and I'll put up the patch for that.

CelticMinstrel commented:
I would say someone wanting the effect of an else-only `[role]` is better off using `[if]`.

GregoryLundberg commented:
Granted. But never underestimate the silliness of programmers. Give 'em a rope for the flag and you'll find your car tied to the trash bins.

ln-zookeeper commented:
The discussion is too scattered to follow, so I'll just ask:

What's the point of search_map=? If you want to limit the search to the recall list only, then x,y=recall,recall is already the one-liner you will (or should) know. Plus, search_map= doesn't allow you to limit the search to the map only, which seems like it'd be a common usecase (find an existing unit to say something, but ignore recallables) for which you'll then have to use x,y= anyway.

As for auto-recalling (I realize it already went in in CelticMinstrel's PR), I find it hard to tell how useful that'd actually be. The role of the [role] tag is pretty straightforward: give a role to a unit. Making it actually able to recall the unit as well feels a bit weird, since that's exactly what [if]s and [recall]s are supposed to be for. Considering the kind of usecases the auto-recalling seems to be intended for, wouldn't it make more sense to replace that with a [then] tag to complement CelticMinstrel's [else]? Then you could have a [role] tag with a [then] that recalls the unit and an [else] that creates a new one if none was found.

GregoryLundberg commented:
@ln-zookeeper I can't speak to why @CelticMinstrel made the original change. From my point of view, we were headed toward needing a set of .. hmm .. I think I was up to 8 or 9 distinct use cases each needing different macros; some quite similar, some quite different.

The search map option I added because the test was there but the variable was always true, so I added a way to set it false. Today's discussion is about it opening the possibility of a do-nothing [role], which I will be fixing after dinner.

Personally, while I see the x,y=recall,recall and can understand it, it seems a hack-y solution. It feels more like an implementation detail. Saying `search_recall_list=only` feels much more literate and less likely to be misunderstood or confused. Consider `x,y=recall,recall search_recall_list=no`

CelticMinstrel commented:
Actually, the reason for the `search_map` variable being there was because I initially implemented `search_recall_list=only` and then retracted it when I realized `x,y=recall,recall` has the same effect.

CelticMinstrel commented on data/lua/wml-tags.lua:
```diff
@@ -911,11 +911,10 @@ function wml_actions.role(cfg)
 
 	filter.role, filter.type = nil, nil
 	local search_map, search_recall, reassign = true, true, true
-	if cfg.search_recall_list ~= nil then
```
Did you test that this works? It's possible that `== "no"` will always be false no matter what. I'd recommend testing for "only" and otherwise using the double-not trick.

GregoryLundberg replied:
Quick test only.

As to 'not not' read http://stackoverflow.com/questions/11806305/why-use-not-not-operation

Which leads me to think there's more I should test ...

GregoryLundberg commented:
OK. All fixed and squashed back to a minimal patch set with updated commit message

I see, now, that the engine converts "yes" to true and "no" to false and passes all other strings through unchanged.

So, if the default is "no" use
    local option = cfg.option
And omitting option= or giving option=no is false; and option=anystring is true

And if the default is "yes" use
    local option = true
    if cfg.option ~= nil then
        option = not not cfg.option
    end
And option=no is false; while omitting the option= or option=anystring is true

And, yes, during my testing I used both "yes" and "anystring" and tested both default yes and default no.

gfgtdf commented:
If you want the default for a boolen attribute to be yes, you can just use `local option = cfg.option ~= false`

CelticMinstrel commented:
This seems fine now. The only remaining issue is whether it will be accepted now (and included in 1.13.5) or whether it will be merged next week. Also, could you add a changelog entry for all the new `[role]` features? You can keep it as a separate commit.

CelticMinstrel commented:
(Actually, thinking about it, there's a small difference between `[auto_recall]` and `[then][recall]` - if there are multiple units with the same role, the latter may recall one even though there was already one on the map, whereas the former presumably would not do that...)

GregoryLundberg commented:
Correct, auto_recall only recalls when (1) reassign=no and the unit was found in the recall list, or (2) the role was assigned to a unit on the recall list (even if there is another with the role also there).

I take it 1.13.5 is being tagged. Are you using the even-stable odd-testing rule? 1.13.4+dev has a lot of UI bugs. The one which is most annoying, today, is that the semi-colon for the command line does not get removed and I keep forgetting to ';' <backspace> 'debug' and get the error ';debug is unknown' or some such.

Personally, I don't care when the PR goes in (or, TBH, if it does), or when .. to me, that's completely up to those who have commit rights.

CelticMinstrel commented:
> The one which is most annoying, today, is that the semi-colon for the command line does not get removed and I keep forgetting to ';' 'debug' and get the error ';debug is unknown' or some such.

Oh yeah, I noticed that too... I'll see if I can fix it in the next few days (or get someone else to).

GregoryLundberg commented:
Let's hold off on merging this PR until 1.13.5+dev. You probably have enough to do with the existing work.

GregoryLundberg commented:
On the question, "Should a role be unique?"

The answer is "HttT depends upon a role NOT being unique."

So, it depends upon the intention of the scenario designer. The [role] tag, however, should ease support for both use cases. I think it does, with the reassign=no option.

CelticMinstrel commented:
Eh? It actually depends on it? Interesting... this makes me curious.

GregoryLundberg commented:
HttT S07 'Crossroads' It uses [random_placement] to scatter ambushes throughout the hills. Each unit it creates has the role 'ambusher' and a filter on move-to tests a moving unit adjacent to a unit of role=ambusher to remove the trait once discovered so, if by-passed, it does not disappear. Nicely done and very efficient using the role like that.

ln-zookeeper commented:
Roles shouldn't need to be unique. Of course most of the time they are, but the engine shouldn't try to enforce it. Whether [role] should let you assign a role that already exists on another unit is a slightly different question though, but I'd be inclined to suggest that it shouldn't care about existing roles except when using reassign=yes or something else necessitating it.

CelticMinstrel commented:
Okay, could you please update this (via `git pull --rebase`) and add a changelog entry?

GregoryLundberg commented:
Rebased and updated. Let me know if there is anything else.

CelticMinstrel commented:
Changelog is slightly wrong (in that search_recall_list was already there, you just added "only" as a possible value), but I'll let this pass.
